timeinterval/timeinterval.go

```diff
@@ -0,0 +1,473 @@
+		return err
+	}
+	err := stringableRangeFromString(str, r)
+	if r.Begin > r.End {
+		return fmt.Errorf("end year %d is before start year %d", r.End, r.Begin)
+	}
+	return err
+}
+
+// UnmarshalYAML implements the Unmarshaller interface for TimeRanges.
+func (tr *TimeRange) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var y yamlTimeRange
+	if err := unmarshal(&y); err != nil {
+		return err
+	}
+	if y.EndTime == "" || y.StartTime == "" {
+		return errors.New("both start and end times must be provided")
+	}
+	start, err := parseTime(y.StartTime)
+	if err != nil {
+		return err
+	}
+	end, err := parseTime(y.EndTime)
+	if err != nil {
+		return err
+	}
+	if start >= end {
+		return errors.New("start time cannot be equal or greater than end time")
+	}
+	tr.StartMinute, tr.EndMinute = start, end
+	return nil
+}
+
+// MarshalYAML implements the yaml.Marshaler interface for WeekdayRange
+func (r WeekdayRange) MarshalYAML() (interface{}, error) {
+	beginStr, ok := daysOfWeekInv[r.Begin]
+	if !ok {
+		return nil, fmt.Errorf("unable to convert %d into weekday string", r.Begin)
+	}
+	if r.Begin == r.End {
+		return interface{}(beginStr), nil
+	}
+	endStr, ok := daysOfWeekInv[r.End]
+	if !ok {
+		return nil, fmt.Errorf("unable to convert %d into weekday string", r.End)
+	}
+	rangeStr := fmt.Sprintf("%s:%s", beginStr, endStr)
+	return interface{}(rangeStr), nil
+}
+
+//MarshalYAML implements the yaml.Marshaler interface for TimeRange
+func (tr TimeRange) MarshalYAML() (out interface{}, err error) {
+	startHr := tr.StartMinute / 60
+	endHr := tr.EndMinute / 60
+	startMin := tr.StartMinute % 60
+	endMin := tr.EndMinute % 60
+
+	startStr := fmt.Sprintf("%02d:%02d", startHr, startMin)
+	endStr := fmt.Sprintf("%02d:%02d", endHr, endMin)
+
+	yTr := yamlTimeRange{startStr, endStr}
+	return interface{}(yTr), err
+}
+
+//MarshalYAML implements the yaml.Marshaler interface for InclusiveRange
+func (ir InclusiveRange) MarshalYAML() (interface{}, error) {
+	if ir.Begin == ir.End {
+		return strconv.Itoa(ir.Begin), nil
+	}
+	out := fmt.Sprintf("%d:%d", ir.Begin, ir.End)
+	return interface{}(out), nil
+}
+
+// TimeLayout specifies the layout to be used in time.Parse() calls for time intervals
+const TimeLayout = "15:04"
+
+var validTime string = "^((([01][0-9])|(2[0-3])):[0-5][0-9])$|(^24:00$)"
+var validTimeRE *regexp.Regexp = regexp.MustCompile(validTime)
+
+// Given a time, determines the number of days in the month that time occurs in.
+func daysInMonth(t time.Time) int {
+	monthStart := time.Date(t.Year(), t.Month(), 1, 0, 0, 0, 0, t.Location())
+	monthEnd := monthStart.AddDate(0, 1, 0)
+	diff := monthEnd.Sub(monthStart)
+	return int(diff.Hours() / 24)
+}
+
+func clamp(n, min, max int) int {
+	if n <= min {
+		return min
+	}
+	if n >= max {
+		return max
+	}
+	return n
+}
+
+// ContainsTime returns true if the TimeInterval contains the given time, otherwise returns false
+func (tp TimeInterval) ContainsTime(t time.Time) bool {
```

Comment:
Do we check somewhere for empty time intervals? WIth an empty time interval, this would always return true

Reply:
While this makes sense due to the semantics of the time interval (where not specifying a restriction means all possible times are included), I think an empty time interval is probably a symptom of an error. I'll add a check for this, thanks.

Reply:
I've thought more about this, I think we should leave this as is. This makes sense with the semantics of a time interval and modifying it would complicate the function by possibly returning only a single error case.